### PR TITLE
[ASan][libc++] Annotating `std::basic_string` with all allocators

### PR DESCRIPTION
**Pull request by AdvenamTacet**
This commit turns on ASan annotations in `std::basic_string` for all allocators by default.

Originally suggested here: https://reviews.llvm.org/D146214

String annotations added here: https://github.com/llvm/llvm-project/pull/72677

This commit is part of our efforts to support container annotations with (almost) every allocator. Annotating `std::basic_string` with default allocator is implemented in https://github.com/llvm/llvm-project/pull/72677.

Additionally it removes `__begin != nullptr` because `data()` should never return a nullptr. 

Support in ASan API exists since https://github.com/llvm/llvm-project/commit/1c5ad6d2c01294a0decde43a88e9c27d7437d157. This patch removes the check in std::basic_string annotation member function (__annotate_contiguous_container) to support different allocators.

 You can turn off annotations for a specific allocator based on changes from https://github.com/llvm/llvm-project/commit/2fa1bec7a20bb23f2e6620085adb257dafaa3be0.

The motivation for a research and those changes was a bug, found by Trail of Bits, in a real code where an out-of-bounds read could happen as two strings were compared via a call to `std::equal` that took `iter1_begin`, `iter1_end`, `iter2_begin` iterators (with a custom comparison function). When object `iter1` was longer than `iter2`, read out-of-bounds on `iter2` could happen. Container sanitization would detect it.

If you have any questions, please email:
- advenam.tacet@trailofbits.com
- disconnect3d@trailofbits.com

### CODE REVIEW

**Comment by AdvenamTacet** [comment]
I think we are ready to turn on annotations for all allocators as no bugs are reported for https://github.com/llvm/llvm-project/pull/72677. @vitalybuka @ldionne let me know if you agree.

Btw. I see that tests are failing, but those fails seem to be unrelated. I'm going to retest  it soon.

File: libcxx/test/libcxx/containers/strings/basic.string/asan_turning_off.pass.cpp
Comment (EricWF) [inline libcxx/test/libcxx/containers/strings/basic.string/asan_turning_off.pass.cpp:58]: I think this encourages users to do bad things with our namespace. 
It also introduces an easy way to have ODR violations.
If we can modify the users allocator type, I would rather do that.
If we can't modify the users allocator type, how do we know that ` std::__asan_annotate_container_with_allocator` will be seen before string with the user allocator?

I think it may also be the wrong default, but time will tell here. I suspect a lot of custom allocators have this behavior.

File: libcxx/test/libcxx/containers/strings/basic.string/asan_turning_off.pass.cpp
Comment (EricWF) [inline libcxx/test/libcxx/containers/strings/basic.string/asan_turning_off.pass.cpp:53]: Please check that the size requested isn't larger than the size of the buffer. Otherwise we have UB.

File: libcxx/test/libcxx/containers/strings/basic.string/asan_turning_off.pass.cpp
Comment (EricWF) [inline libcxx/test/libcxx/containers/strings/basic.string/asan_turning_off.pass.cpp:15]: Allocators don't destroy things internally, but the standard library may use an allocator to call a destructor. 

Does this mean to say "Some allocators don't call free from deallocate?" which would leave the memory in the  state string left it in?

If we really do mean "There's code that reuses the memory of live objects without destroying them", then that's exactly the kind of bug ASAN is meant to catch. That is not UB we should accept.

File: libcxx/test/libcxx/containers/strings/basic.string/asan_turning_off.pass.cpp
Comment (EricWF) [inline libcxx/test/libcxx/containers/strings/basic.string/asan_turning_off.pass.cpp:66]: I'm not sure I understand what this test is doing.

It looks like it intends to get the reuse_buffers to all be used, but `push_back` doesn't allocate a new buffer on each call. But it only likely consumes 3-4 of the 100 buffers.

Then we reset (makes sense) and using the same memory we clobber `S`, then `S`'s old buffers, then S's current buffer. Which is all very much UB.

File: libcxx/include/string
Comment (EricWF) [inline libcxx/include/string:1894]: Why did the `__begin` != nullptr get removed?

File: libcxx/test/libcxx/containers/strings/basic.string/asan.pass.cpp
Comment (EricWF) [inline libcxx/test/libcxx/containers/strings/basic.string/asan.pass.cpp:50]: `volatile` on `T foo`  won't prevent the compiler from optimizing away the read from `std::string`. If the compiler can constant fold `foo`'s initializer, it doesn't have to perform the ASAN triggering memory access.

Comment (EricWF) [inline libcxx/test/libcxx/containers/strings/basic.string/asan.pass.cpp:50]: Won't hardening mode catch the out-of-bounds access too?

**Review by EricWF — CHANGES_REQUESTED** [review]
I'm pretty opposed to a couple of aspects of this patch.

1. I think the customization point `std::__asan_annotate_container_with_allocator` encourages users to (A) use internals, (B) specialize STL types, and (C) configure the customization point for types they don't own.

2. The added tests contain a bunch of UB. 


Could you please provide an example that is well-formed? Then we can figure out a more appropriate way to provide the customization needed.

File: libcxx/test/libcxx/containers/strings/basic.string/asan_turning_off.pass.cpp
Comment (AdvenamTacet) [reply libcxx/test/libcxx/containers/strings/basic.string/asan_turning_off.pass.cpp:58]: It already works in `std::deque` and `std::vector` case and I'm not aware of any reported issues.

File: libcxx/include/string
Comment (AdvenamTacet) [reply libcxx/include/string:1894]: I should probably add it in the description of the PR, but `data()` cannot be nullptr, so I'm removing unnecessary check.

File: libcxx/test/libcxx/containers/strings/basic.string/asan_turning_off.pass.cpp
Comment (AdvenamTacet) [reply libcxx/test/libcxx/containers/strings/basic.string/asan_turning_off.pass.cpp:15]: This test case check a solution to the only known, potential issue with annotating all allocators, situations when memory is repurposed and destructors are not called.
Related issue: https://github.com/llvm/llvm-project/issues/60384

That's the only issue from past, I'm aware of (except bugs). That was addressed here: https://reviews.llvm.org/D145628
And it successfully works in `std::vector` and `std::deque`.

But is it UB? I remember that it's not clear.
Related discussion: https://reviews.llvm.org/D136765#4155262

As pointed out by @AntonBikineev 
> A program may end the lifetime of an object of class type without invoking the destructor, by reusing or releasing the storage as described above[.](https://eel.is/c++draft/basic.life#5.sentence-1)
> [...]
> [Note [4](https://eel.is/c++draft/basic.life#note-4): The correct behavior of a program often depends on the destructor being invoked for each object of class type[.](https://eel.is/c++draft/basic.life#5.sentence-4)
— end note]

Source:  https://eel.is/c++draft/basic.life#6

Therefore, it's not clear if it's UB. But even if it is, we want to make sure that it works that way, because people rely on this behavior.

File: libcxx/test/libcxx/containers/strings/basic.string/asan_turning_off.pass.cpp
Comment (AdvenamTacet) [reply libcxx/test/libcxx/containers/strings/basic.string/asan_turning_off.pass.cpp:66]: I answered under your another, related comment: https://github.com/llvm/llvm-project/pull/75845/files/d297670ad62f75d64c8fbef867d2e487390a4a1d#r1431944505

File: libcxx/test/libcxx/containers/strings/basic.string/asan.pass.cpp
Comment (AdvenamTacet) [reply libcxx/test/libcxx/containers/strings/basic.string/asan.pass.cpp:50]: In this very code, if it's enabled, yes. But if we use a pointer to access that memory (or C function taking pointers etc.) then no.

File: libcxx/include/string
Comment (AdvenamTacet) [reply libcxx/include/string:1894]: In your example, the string is short (btw. the default allocator is also used), and until https://github.com/llvm/llvm-project/pull/75882 annotations for SSO cases are turned off. But even if you turn them on (change value of `_LIBCPP_SHORT_STRING_ANNOTATIONS_ALLOWED` to true) it works as expected. Constant evaluated objects are not annotated, runtime objects are annotated. s2 is annotated when it's created.

s is probably not annotated, but it's ok, it may only lead to false negatives. I will investigate it before merge of short string annotations, but it shouldn't affect this PR.

No boom is expected. No boom happens.

**Review by EricWF — COMMENTED** [review]
OK, I have a few requested changes. Then I say we land this, and then take a short break to ensure everything that's landed so far is going to stick, and to do some cleanup.

One cleanup we need to undertake is one that removes our dependence on the optimizer removing dead calls for us (even if it does, that's likely at the cost of some other code it can't optimize because it's hit its inliner limits).

I'm going to hold off actually approving this patch until we figure out if the performance fixes for the earlier patches are sufficient.

File: libcxx/test/libcxx/containers/strings/basic.string/asan.pass.cpp
Comment (AdvenamTacet) [reply libcxx/test/libcxx/containers/strings/basic.string/asan.pass.cpp:50]: Good point, however this test runs only when ASan is present (`// REQUIRES: asan`). If we ever create test for ASan and hardening mode at the same time, we have to exclude that test from it.

File: libcxx/test/libcxx/containers/strings/basic.string/asan_turning_off.pass.cpp
Comment (AdvenamTacet) [reply libcxx/test/libcxx/containers/strings/basic.string/asan_turning_off.pass.cpp:58]: Yes, there are users of it as it was added for a direct request.

I think https://github.com/llvm/llvm-project/pull/75845#discussion_r1431944505 is a continuation of that discussion, so I resolve this conversation.

File: libcxx/include/string
Comment (AdvenamTacet) [reply libcxx/include/string:1894]: It can tolerate unpoisoned memory where it should be poisoned, not the other way around. It may result in some false negatives, but we accept that.

File: libcxx/test/libcxx/containers/strings/basic.string/asan.pass.cpp
Comment (AdvenamTacet) [reply libcxx/test/libcxx/containers/strings/basic.string/asan.pass.cpp:50]: Afair, ASan checks are added before optimizations, so volatile is not necessary, but it will be checked.

File: libcxx/test/libcxx/containers/strings/basic.string/asan_turning_off.pass.cpp
Comment (AdvenamTacet) [reply libcxx/test/libcxx/containers/strings/basic.string/asan_turning_off.pass.cpp:15]: Added comments.

**Review by EricWF — APPROVED** [review]
I think this LGTM, but it needs updating since we're not going to land the `if constexpr` magic it's using. But I think we have a good handle on ensuring this is no overhead by other means.

If you wouldn't mind giving me a few days to test this internally before landing, I would appreciate it. But if there's time pressure to land this, that's OK too.

**Comment by AdvenamTacet** [comment]
I will remove the constexpr magic Tomorrow.

No big time pressure, I believe we can wait a few days. Probably it's good to test before landing to avoid reverting. Also, @ldionne didn't stamp this PR yet.

Would be good, however, to work on short string annotations in parallel. I will update/finish https://github.com/llvm/llvm-project/pull/75882 Tomorrow, could you look at it as well then?

File: libcxx/test/libcxx/containers/strings/basic.string/asan_turning_off.pass.cpp
Comment (frederick-vs-ja) [reply libcxx/test/libcxx/containers/strings/basic.string/asan_turning_off.pass.cpp:15]: > If we really do mean "There's code that reuses the memory of live objects without destroying them", then that's exactly the kind of bug ASAN is meant to catch. That is not UB we should accept.

This might by considered as UB in the past, but is not UB now per [CWG2523](https://cplusplus.github.io/CWG/issues/2523.html).

Comment (AdvenamTacet) [reply libcxx/test/libcxx/containers/strings/basic.string/asan_turning_off.pass.cpp:15]: Thx for the link! We do take care of that case, so no changes required, but I will add that link to the comment in the test.

**Review by ldionne — APPROVED** [review]

**Comment by AdvenamTacet** [comment]
@ldionne  Thx for looking at it! I will merge when CI ends running.

**Comment by AdvenamTacet** [comment]
Buildbots are failing. https://lab.llvm.org/buildbot/#/builders/37/builds/29906

[Error](https://lab.llvm.org/buildbot/#/builders/37/builds/29906/steps/13/logs/stdio) is not clear for me, but I think problem is not in that PR, therefore I'm not reverting.

```
/b/sanitizer-x86_64-linux/build/llvm-project/compiler-rt/test/fuzzer/fork.test:15:8: error: CRASH: expected string not found in input
CRASH: {{SEGV|access-violation}} on unknown address 0x00000000
       ^
```

Could it be that another error was detected?

**Pull request merged**
